File: Sentence Similarity III.py
class Solution:
    def areSentencesSimilar(self, sentence1: str, sentence2: str) -> bool:
        splits1 = sentence1.split()
        splits2 = sentence2.split()
        if len(splits1) > len(splits2):
            splits1, splits2 = splits2, splits1

        n1, n2 = len(splits1), len(splits2)
        if n1 == 1:
            return splits2[0] == splits1[0] or splits2[-1] == splits1[0]

        l1, r1 = 0, n1 - 1
        r2 = n2 - 1
        while l1 < n1 and splits1[l1] == splits2[l1]:
            l1 += 1
        while r1 >= l1 and splits1[r1] == splits2[r2]:
            r1 -= 1
            r2 -= 1
        # print(l1, r1, r2)
        return r1 < l1

File: Sentence Similarity III.cpp
class Solution {
public:
    vector<string> split(string& s) {
        stringstream ss(s);
        vector<string> v;
        while (getline(ss, s, ' ')) {
            v.push_back(s);
        }
        return v;
    }

    bool areSentencesSimilar(string sentence1, string sentence2) {
        auto splits1 = split(sentence1), splits2 = split(sentence2);
        if (splits1.size() > splits2.size()) swap(splits1, splits2);
        int n1 = splits1.size(), n2 = splits2.size();
        if (n1 == 1) return (splits2[0] == splits1[0] || splits2[n2-1] == splits1[0]);

        int l1 = 0, r1 = n1 - 1, r2 = n2 - 1;
        while (l1 < n1 && splits1[l1] == splits2[l1]) l1++;
        while (r1 >= l1 && splits1[r1] == splits2[r2]) {
            r1--;
            r2--;
        }
        return r1 <l1;
    }
};
